### PR TITLE
feat: Add BM25 and Hybrid Search Retrievers to Azure AI Search Integration

### LINKED ISSUE #950: Integration for Azure AI Search
## Summary and motivation

Motivation from this issue [#6749](https://github.com/deepset-ai/haystack/issues/6749) raised by community members 
## Detailed design

Create a new document store for Azure AI search allowing operations of vector search, hybrid search and keyword search.

## Implementation Breakdown

- [x] Create a document store with CURD operations.
- [x]  Tests for document stores.
- [ ]  Add retrievers.
- [ ] Tests for retrievers.
- [x] Add filters and search operations.
- [ ] Test for filters.

## Checklist

If the request is accepted, ensure the following checklist is complete before closing this issue.
```[tasklist]
### Tasks
- [ ] The code is documented with docstrings and was merged in the `main` branch
- [ ] Docs are published at https://docs.haystack.deepset.ai/
- [ ] There is a Github workflow running the tests for the integration nightly and at every PR
- [ ] A label named like `integration:<your integration name>` has been added to this repo
- [ ] The [labeler.yml](https://github.com/deepset-ai/haystack-core-integrations/blob/main/.github/labeler.yml) file has been updated
- [ ] The package has been released on PyPI
- [ ] An integration tile has been added to https://github.com/deepset-ai/haystack-integrations
- [ ] The integration has been listed in the [Inventory section](https://github.com/deepset-ai/haystack-core-integrations#inventory) of this repo README
- [ ] There is an example available to demonstrate the feature
- [ ] The feature was announced through social media
```

### PR DESCRIPTION
**Pull request by ttmenezes**
### Related Issues

- fixes #950 

### Proposed Changes:

 
 
This PR includes the following features:
- Implement a BM25 retriever for Azure AI Search
- Implement a hybrid search retriever for Azure AI Search

### How did you test it?

- Manual testing in Haystack pipelines
- Added unit tests, using existing pattern from embedding retriever tests

### Notes for the reviewer

Used the existing embedding retriever code as a template and primarily just changed the parameters in the calls to the Azure AI Search Python SDK.

### Checklist

- I have read the [contributors guidelines](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CODE_OF_CONDUCT.md)
- I have updated the related issue with new insights and changes
- I added unit tests and updated the docstrings
- I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.

### CODE REVIEW

File: integrations/azure_ai_search/src/haystack_integrations/components/retrievers/azure_ai_search/bm25_retriever.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+    """
+
+    def __init__(
+        self,
+        *,
+        document_store: AzureAISearchDocumentStore,
+        filters: Optional[Dict[str, Any]] = None,
+        top_k: int = 10,
+        filter_policy: Union[str, FilterPolicy] = FilterPolicy.REPLACE,
```
Comment (Amnah199) [inline integrations/azure_ai_search/src/haystack_integrations/components/retrievers/azure_ai_search/bm25_retriever.py:27]: If we want to enable semantic ranking in all retrievers, we can add `**kwargs` here too.
```suggestion
        filter_policy: Union[str, FilterPolicy] = FilterPolicy.REPLACE,
        **kwargs,
```

File: integrations/azure_ai_search/src/haystack_integrations/components/retrievers/azure_ai_search/bm25_retriever.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    ):
+        """
+        Create the AzureAISearchBM25Retriever component.
+
+        :param document_store: An instance of AzureAISearchDocumentStore to use with the Retriever.
+        :param filters: Filters applied when fetching documents from the Document Store.
+            Filters are applied during the BM25 search to ensure the Retriever returns
+              `top_k` matching documents.
+        :param top_k: Maximum number of documents to return.
+        :filter_policy: Policy to determine how filters are applied. Possible options:
```
Comment (Amnah199) [inline integrations/azure_ai_search/src/haystack_integrations/components/retrievers/azure_ai_search/bm25_retriever.py:37]:
```suggestion
        :filter_policy: Policy to determine how filters are applied.
        :param kwargs: Additional keyword arguments to pass to the Azure AI's search endpoint.
            Some of the supported parameters:
                - `query_type`: A string indicating the type of query to perform. Possible values are
                'simple','full' and 'semantic'.
                - `semantic_configuration_name`: The name of semantic configuration to be used when
                processing semantic queries.
```

**Review by Amnah199 — APPROVED** [review]
Can merge this with one more approval.

**Review by julian-risch — CHANGES_REQUESTED** [review]
Looks quite good to me already. I have a couple of very small change requests. Another suggestion for improvement would be to cover the run method with a unit test and mock for example _bm25_retrieval. The test could check that _bm25_retrieval is called with the expected parameters. Similarly for the hybrid retriever.

File: integrations/azure_ai_search/src/haystack_integrations/components/retrievers/azure_ai_search/hybrid_retriever.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    ):
+        """
+        Create the AzureAISearchHybridRetriever component.
+
+        :param document_store: An instance of AzureAISearchDocumentStore to use with the Retriever.
+        :param filters: Filters applied when fetching documents from the Document Store.
+            Filters are applied during the hybrid search to ensure the Retriever returns
+              `top_k` matching documents.
+        :param top_k: Maximum number of documents to return.
+        :filter_policy: Policy to determine how filters are applied.
```
[the remark this answers is not in the document]
Comment (ttmenezes) [reply integrations/azure_ai_search/src/haystack_integrations/components/retrievers/azure_ai_search/hybrid_retriever.py:38]: Added.

File: integrations/azure_ai_search/src/haystack_integrations/components/retrievers/azure_ai_search/bm25_retriever.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    ):
+        """
+        Create the AzureAISearchBM25Retriever component.
+
+        :param document_store: An instance of AzureAISearchDocumentStore to use with the Retriever.
+        :param filters: Filters applied when fetching documents from the Document Store.
+            Filters are applied during the BM25 search to ensure the Retriever returns
+              `top_k` matching documents.
+        :param top_k: Maximum number of documents to return.
+        :filter_policy: Policy to determine how filters are applied.
```
[the remark this answers is not in the document]
Comment (ttmenezes) [reply integrations/azure_ai_search/src/haystack_integrations/components/retrievers/azure_ai_search/bm25_retriever.py]: Added.

File: integrations/azure_ai_search/src/haystack_integrations/components/retrievers/azure_ai_search/bm25_retriever.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        :param kwargs: Additional keyword arguments to pass to the Azure AI's search endpoint.
+            Some of the supported parameters:
+                - `query_type`: A string indicating the type of query to perform. Possible values are
+                'simple','full' and 'semantic'.
+                - `semantic_configuration_name`: The name of semantic configuration to be used when
+                processing semantic queries.
+            For more information on parameters, see the
+            [official Azure AI Search documentation](https://learn.microsoft.com/en-us/azure/search/).
```
[the remark this answers is not in the document]
Comment (ttmenezes) [reply integrations/azure_ai_search/src/haystack_integrations/components/retrievers/azure_ai_search/bm25_retriever.py:46]: Added.

File: integrations/azure_ai_search/src/haystack_integrations/components/retrievers/azure_ai_search/hybrid_retriever.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        :param kwargs: Additional keyword arguments to pass to the Azure AI's search endpoint.
+            Some of the supported parameters:
+                - `query_type`: A string indicating the type of query to perform. Possible values are
+                'simple','full' and 'semantic'.
+                - `semantic_configuration_name`: The name of semantic configuration to be used when
+                processing semantic queries.
+            For more information on parameters, see the
+            [official Azure AI Search documentation](https://learn.microsoft.com/en-us/azure/search/).
```
[the remark this answers is not in the document]
Comment (ttmenezes) [reply integrations/azure_ai_search/src/haystack_integrations/components/retrievers/azure_ai_search/hybrid_retriever.py:46]: Added.

File: integrations/azure_ai_search/src/haystack_integrations/document_stores/azure_ai_search/document_store.py [hunk 7]
```diff
@@ -421,21 +421,20 @@ def _embedding_retrieval(
...
-        :param filters: Filters applied to the retrieved Documents. Defaults to None.
-            Filters are applied during the approximate kNN search to ensure that top_k matching documents are returned.
+        :param top_k: Maximum number of Documents to return.
+        :param filters: Filters applied to the retrieved Documents.
         :param kwargs: Optional keyword arguments to pass to the Azure AI's search endpoint.
 
-        :raises ValueError: If `query_embedding` is an empty list
-        :returns: List of Document that are most similar to `query_embedding`
+        :raises ValueError: If `query_embedding` is an empty list.
+        :returns: List of Document that are most similar to `query_embedding`.
```
[the remark this answers is not in the document]
Comment (ttmenezes) [reply integrations/azure_ai_search/src/haystack_integrations/document_stores/azure_ai_search/document_store.py:440]: Added.

File: integrations/azure_ai_search/src/haystack_integrations/components/retrievers/azure_ai_search/hybrid_retriever.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        top_k: Optional[int] = None,
+    ):
+        """Retrieve documents from the AzureAISearchDocumentStore.
+
+        :param query: Text of the query.
+        :param query_embedding: A list of floats representing the query embedding
+        :param filters: Filters applied to the retrieved Documents. The way runtime filters are applied depends on
+                        the `filter_policy` chosen at retriever initialization. See `__init__` method docstring for more
+                        details.
+        :param top_k: The maximum number of documents to retrieve.
```
[the remark this answers is not in the document]
Comment (ttmenezes) [reply integrations/azure_ai_search/src/haystack_integrations/components/retrievers/azure_ai_search/hybrid_retriever.py:115]: Added.

File: integrations/azure_ai_search/src/haystack_integrations/components/retrievers/azure_ai_search/bm25_retriever.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    @component.output_types(documents=List[Document])
+    def run(self, query: str, filters: Optional[Dict[str, Any]] = None, top_k: Optional[int] = None):
+        """Retrieve documents from the AzureAISearchDocumentStore.
+
+        :param query: Text of the query.
+        :param filters: Filters applied to the retrieved Documents. The way runtime filters are applied depends on
+                        the `filter_policy` chosen at retriever initialization. See init method docstring for more
+                        details.
+        :param top_k: the maximum number of documents to retrieve.
+        :returns: a dictionary with the following keys:
```
[the remark this answers is not in the document]
Comment (ttmenezes) [reply integrations/azure_ai_search/src/haystack_integrations/components/retrievers/azure_ai_search/bm25_retriever.py:109]: Added.

**Review by julian-risch — APPROVED** [review]
Looks good to me! 👍 
Edit: I talked to @Amnah199 offline about adding unit tests for the run methods.

**Pull request merged**